# 5 УРОК

ВРЕМЯ И СВОДНЫЕ ТАБЛИЦЫ

Имеются следующие данные о транзакциях в период с 01.12.2010 по 12.09.2011:

InvoiceNo — номер транзакции
StockCode — код товара
Description — описание товара
Quantity — количество единиц товара, добавленных в заказ
InvoiceDate — дата транзакции 
UnitPrice — цена за единицу товара
CustomerID — id клиента
Country — страна, где проживает клиент
Данные можно скачать отсюда.

В данной задаче ссылка для считывания данных лежит в переменной path_to_file.
Импортируйте пандас и прочитайте данные с кодировкой ISO-8859-1. Запишите полученный датафрейм в retail, а названия колонок сохраните в переменную retail_columns.

In [7]:
path_to_file=('/mnt/HC_Volume_18315164/home-jupyter/jupyter-s-kamsarin/learn_pandas/data.csv')
import pandas as pd
retail = pd.read_csv(path_to_file, encoding='ISO-8859-1', parse_dates=['InvoiceDate'])
retail_columns = retail.columns

In [8]:
retail.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [9]:
retail_columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

In [10]:
retail.dtypes

InvoiceNo              object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[ns]
UnitPrice             float64
CustomerID            float64
Country                object
dtype: object

In [12]:
retail.shape

(541909, 8)

In [13]:
retail.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


Задание 2
Проверьте, встречаются ли в данных повторяющиеся наблюдения, и в качестве ответа укажите их количество.
Если они есть, то удалите их из retail.

In [14]:
retail.duplicated().sum() # Можно и так было сделать

5268

In [18]:
retail.shape[0]


541909

In [15]:
start_rows = retail.shape[0]

In [19]:
retail = retail.drop_duplicates()

In [20]:
start_rows - retail.shape[0]

5268

⭐️ Задание со звёздочкой! ⭐️

Данные содержат в себе записи как об успешных транзакциях, так и об отмененных.
Если пользователь отменил заказ, в начале номера транзакции (InvoiceNo) ставится C (canceled). 

Сколько всего транзакций отменили пользователи? Считайте, что каждая строка - это отдельная транзакция.

In [22]:
retail

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
...                    ...        ...         ...             ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France  
541905 2011-12-09 12:50:00       2.10     12680.0          France  
541906 2011-12-09 12:50:00       4.15     12680.0          France  
541907 2011-12-09 12:50:00       4.15     12680.0          France  
541908 2011-12-09 12:50:00       4.95     12680.0          France  

[536641 rows x 8 columns]

In [25]:
break_tran = retail.InvoiceNo.str.startswith('C').sum()

In [26]:
break_tran

9251

In [28]:
# можно их и поглядеть ))
retail[retail.InvoiceNo.str.startswith('C')]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

               InvoiceDate  UnitPrice  CustomerID         Country  
141    2010-12-01 09:41:00      27.50     14527.0  United Kingdom  
154    2010-12-01 09:49:00       4.65     15311.0  United Kingdom  
235    2010-12-01 10:24:00       1.65     17548.0  United Kingdom  
236    2010-12-01 10:24:00       0.29     17548.0  United Kingdom  
237    2010-12-01 10:24:00       0.29     17548.0  United Kingdom  
...                    ...        ...         ...             ...  
540449 2011-12-09 09:57:00       0.83     14397.0  United Kingdom  
541541 2011-12-09 10:28:00     224.69     15498.0  United Kingdom  
541715 2011-12-09 11:57:00      10.95     15311.0  United Kingdom  
541716 2011-12-09 11:58:00       1.25     17315.0  United Kingdom  
541717 2011-12-09 11:58:00       1.25     17315.0  United Kingdom  

[9251 rows x 8 columns]

ЗАДАНИЕ
Теперь отфильтруйте данные и оставьте в retail только те заказы, где Quantity > 0. 
В качестве ответа укажите число оставшихся строк.

In [27]:
retail.query('Quantity > 0')

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
...                    ...        ...         ...             ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France  
541905 2011-12-09 12:50:00       2.10     12680.0          France  
541906 2011-12-09 12:50:00       4.15     12680.0          France  
541907 2011-12-09 12:50:00       4.15     12680.0          France  
541908 2011-12-09 12:50:00       4.95     12680.0          France  

[526054 rows x 8 columns]

In [29]:
retail = retail.query('Quantity > 0')

In [30]:
retail.shape[0]

526054

💀 Сложное задание! 💀

Посчитайте число заказов для каждого пользователя (CustomerID) из Германии (Germany). Оставьте только тех, 
кто совершил более N транзакций (InvoiceNo), где N – 80% процентиль. Запишите полученные id пользователей в germany_top 
(не весь датафрейм, только id).

Отфильтрованные данные записаны в retail. Идентификатор заказа – InvoiceNo. 
Для каждого заказа в данных может встречаться более 1 строки.

In [43]:
quantile80 = int(retail.query('Country == "Germany"')\
    .groupby('CustomerID')\
    .agg({'InvoiceNo' : 'nunique'})\
    .quantile(0.8))

In [44]:
quantile80

7

In [50]:
retail.query('Country == "Germany"')\
    .groupby('CustomerID')\
    .agg({'InvoiceNo' : 'nunique'})\
    .rename(columns={'InvoiceNo': 'InvoiceCount'})\
    .query("InvoiceCount > @quantile80")

InvoiceCount
CustomerID              
12471.0               30
12474.0               18
12476.0               11
12481.0               10
12500.0               11
12524.0                8
12569.0               32
12600.0               10
12619.0                9
12621.0               20
12626.0                9
12647.0               12
12662.0               11
12705.0                9
12708.0               10
12709.0               15
12712.0               13
12720.0               25

In [47]:
germany_top = retail.query('Country == "Germany"')\
    .groupby('CustomerID')\
    .agg({'InvoiceNo' : 'nunique'})\
    .rename(columns={'InvoiceNo': 'InvoiceCount'})\
    .query("InvoiceCount > @quantile80")\
    .index


In [48]:
germany_top

Float64Index([12471.0, 12474.0, 12476.0, 12481.0, 12500.0, 12524.0, 12569.0,
              12600.0, 12619.0, 12621.0, 12626.0, 12647.0, 12662.0, 12705.0,
              12708.0, 12709.0, 12712.0, 12720.0],
             dtype='float64', name='CustomerID')

⭐️ Задание со звёздочкой! ⭐️

Используя объект с id пользователей (germany_top), полученный на предыдущем шаге,
отфильтруйте наблюдения и оставьте в данных записи только по интересующим нас юзерам. Результирующий датафрейм запишите в top_retail_germany.

Датафрейм записан в retail, нужные id пользователей – в germany_top

In [53]:
top_retail_germany = retail.query('CustomerID in @germany_top')

In [54]:
top_retail_germany

InvoiceNo StockCode                          Description  Quantity  \
1109      536527     22809              SET OF 6 T-LIGHTS SANTA         6   
1110      536527     84347  ROTATING SILVER ANGELS T-LIGHT HLDR         6   
1111      536527     84945   MULTI COLOUR SILVER T-LIGHT HOLDER        12   
1112      536527     22242        5 HOOK HANGER MAGIC TOADSTOOL        12   
1113      536527     22244           3 HOOK HANGER MAGIC GARDEN        12   
...          ...       ...                                  ...       ...   
541726    581570     22139     RETROSPOT TEA SET CERAMIC 11 PC          3   
541727    581570     23077                  DOUGHNUT LIP GLOSS         20   
541728    581570     20750             RED RETROSPOT MINI CASES         2   
541729    581570     22505            MEMO BOARD COTTAGE DESIGN         4   
541730    581570      POST                              POSTAGE         1   

               InvoiceDate  UnitPrice  CustomerID  Country  
1109   2010-12-01 13:04:00       2.95     12662.0  Germany  
1110   2010-12-01 13:04:00       2.55     12662.0  Germany  
1111   2010-12-01 13:04:00       0.85     12662.0  Germany  
1112   2010-12-01 13:04:00       1.65     12662.0  Germany  
1113   2010-12-01 13:04:00       1.95     12662.0  Germany  
...                    ...        ...         ...      ...  
541726 2011-12-09 11:59:00       4.95     12662.0  Germany  
541727 2011-12-09 11:59:00       1.25     12662.0  Germany  
541728 2011-12-09 11:59:00       7.95     12662.0  Germany  
541729 2011-12-09 11:59:00       4.95     12662.0  Germany  
541730 2011-12-09 11:59:00      18.00     12662.0  Germany  

[4175 rows x 8 columns]

Сгруппируйте top_retail_germany по коду товара (StockCode). Какой из продуктов добавляли в корзину чаще всего, кроме POST? 

Note: одним заказом считается единовременная покупка любого количества товара, т.е. без учета Quantity.

Подсказки

In [61]:
top_retail_germany.groupby('StockCode', as_index = False)\
    .agg({'Quantity':'count'})\
    .rename(columns={'Quantity': 'QuantityCount'})\
    .sort_values('QuantityCount')

StockCode  QuantityCount
0        10125              1
388      22246              1
804      23179              1
385      22242              1
807      23184              1
...        ...            ...
45       20719             30
453      22423             34
411      22328             38
409      22326             52
1157      POST            213

[1158 rows x 2 columns]

Вернемся к анализу полного датасета retail. Создайте колонку Revenue с суммой покупки, 
используя колонки Quantity и UnitPrice.

Отфильтрованные данные записаны в retail.

In [63]:
retail.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [64]:
retail['Revenue'] = retail['Quantity'] * retail['UnitPrice']

/opt/tljh/user/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [65]:
retail.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  Revenue  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom    15.30  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom    20.34  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom    22.00  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom    20.34  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom    20.34

Для каждой транзакции (InvoiceNo), посчитайте финальную сумму заказа. В качестве ответа укажите топ-5 (InvoiceNo) по сумме заказа (через запятую с пробелом и в порядке убывания TotalRevenue)

Например, для следующего примера...

InvoiceNo StockCode Quantity InvoiceDate          UnitPrice  CustomerID  Revenue
536365    85123A    6        2010-12-01 08:26:00  2.55       17850.0     15.30
536365    71053     6        2010-12-01 08:26:00  3.39       17850.0     20.34
555555    71053     1        2010-12-02 18:00:00  3.39       17850.0     3.39


...сумма заказа будет равна:

InvoiceNo  TotalRevenue
536365     35.64
555555     3.39


И ответом будет:



In [72]:
result = retail.groupby('InvoiceNo').agg({'Revenue':'sum'}).sort_values('Revenue', ascending = False).head(5).index.to_list()


In [76]:
print(*result, sep =', ')

581483, 541431, 574941, 576365, 556444
